AccessCheckoutDemo/AccessCheckoutDemoUITests/CardFlowValidationTests.swift

```diff
@@ -243,4 +243,96 @@ class CardFlowCardValidationTests: XCTestCase {
         // Invalid state
         XCTAssertFalse(view!.submitButton.isEnabled)
     }
+    
+    // Mark: Formatting
+    
+    func testCorrectlyFormatsAmexPan() {
+        view!.typeTextIntoPan("37178")
+        XCTAssertTrue(view!.imageIs("amex"))
+        
+        XCTAssertEqual(view!.panText!, "3717 8")
+        
+        view!.typeTextIntoPan("1111")
+        
+        XCTAssertEqual(view!.panText!, "3717 81111")
+        
+        view!.typeTextIntoPan("1111")
+        
+        XCTAssertEqual(view!.panText!, "3717 811111 111")
+        
+        view!.typeTextIntoPan(backspace)
+        view!.typeTextIntoPan(backspace)
+        view!.typeTextIntoPan(backspace)
+        view!.typeTextIntoPan(backspace)
+        
+        XCTAssertEqual(view!.panText!, "3717 81111")
+    }
+    
+    func testCorrectlyFormatsUnknownPan() {
+        view!.typeTextIntoPan("11111")
+        XCTAssertTrue(view!.imageIs("unknown_card_brand"))
+        
+        XCTAssertEqual(view!.panText!, "1111 1")
+        
+        view!.typeTextIntoPan("1111")
+        
+        XCTAssertEqual(view!.panText!, "1111 1111 1")
+        
+        view!.typeTextIntoPan("3333")
+        
+        XCTAssertEqual(view!.panText!, "1111 1111 1333 3")
+        
+        view!.typeTextIntoPan(backspace)
+        view!.typeTextIntoPan(backspace)
+        
+        XCTAssertEqual(view!.panText!, "1111 1111 133")
+    }
+    
+    func testCorrectlyFormatsVisaPan() {
+        view!.typeTextIntoPan("41111")
+        XCTAssertTrue(view!.imageIs("visa"))
+        
+        XCTAssertEqual(view!.panText!, "4111 1")
+        
+        view!.typeTextIntoPan("1111")
+        
+        XCTAssertEqual(view!.panText!, "4111 1111 1")
+        
+        view!.typeTextIntoPan("3333")
+        
+        XCTAssertEqual(view!.panText!, "4111 1111 1333 3")
+        
+        view!.typeTextIntoPan(backspace)
+        view!.typeTextIntoPan(backspace)
+        
+        XCTAssertEqual(view!.panText!, "4111 1111 133")
+    }
+    
+    func testCorrectlyFormatsWhenTypingInMiddleOfPan() {
+        view!.typeTextIntoPan("4111")
+        view!.setCursorAndTypeInPanField(["4"])
+        
+        XCTAssertEqual(view!.panText!, "4114 1")
+    }
+    
+    func testCorrectlyFormatsWhenPastingAndTypingInMiddleOfPan() {
+        view!.typeTextIntoPan("4111")
+        view!.setCursorAndTypeInPanField(["123", "5"])
+        
+        XCTAssertEqual(view!.panText!, "4111 2351")
+    }
+    
+    func testCorrectlyFormatsWhenDeletingInMiddleOfPan() {
+        view!.typeTextIntoPan("4321 4321")
+        view!.setCursorAndTypeInPanField([backspace])
+        
+        XCTAssertEqual(view!.panText!, "4314 321")
+    }
+    
+    func testCorrectlyFormatsAndCursorIsCorrectWhenDeletingAndThenTypingInMiddleOfPan() {
```

Comment:
What do we mean by "AndCursorIsCorrect" since we don't assert on the caret position? Is there an assert missing?

Reply:
I am 'asserting' on the cursor position by asserting on the final text and where that means the typing occurred after the pasting and deleting